### PR TITLE
[Intel GPU] xpu-ops codegen via backend whitelist

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Motivation

This PR intends to enhance the codegen to allow generate codes for XPU backend.

XPU operators need be registered in an hand-written way currently. Developers have no chance to take the advantage of shared code to handle tensor meta setting (like strides, proxy output, structured kernels).  Manually porting code is erro-prone and may lead to high maintaining efforts.

We utilize the backend_whitelist argument in `gen.py` to generate XPU needed headers and source codes.

# Usage
XPU ops lie in `third_pary/torch-xpu-ops`, the codegen process is triggered before the complation of `torch-xpu-ops`

We use the following commands to generate XPU operators

` python -m torchgen.gen --source-path path/to/yaml/of/xpu   --install-dir  build/xpu    --per-operator-headers    --static-dispatch-backend     --backend-whitelist=XPU`

The diff lies at `backend-whitelist=XPU`.  The backend-whitelist key is an existent argument in torchgen. 

The input of `gen.py` are code templates and operators yaml. We share the same templates in `aten`. A simplified yaml lies in `third_party/torch-xpu-ops`, which only includes the supported xpu operators. This yaml is a copy-and-modify of `native_functions.yaml`. No extra entry is added, the format is same as the one in `aten`

# Result

All operators headers are generated in `build/xpu/ATen/ops` independently, which would not affect operators declared/defined by CPU/CUDA or any other backend.  XPU operators only include headers in this folder.

# Verification

* In `third-party/torch-xpu-ops`, we migrate all supported kernels to structured kernels style, where they are registered through `REGISTER_XPU_DISPATCH` or `TORCH_IMPL_FUNC`, and we have UT verification based on `test_ops.py`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130088
* __->__ #130082
* #130019

cc @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

File: tools/test/test_codegen.py [hunk 1]
```diff
@@ -498,6 +511,164 @@ def test_op_with_cpp_sig_generates_static_dispatch(self) -> None:
...
+            rocm=False,
+            per_operator_headers=True,
+        )
+
+    def file_has_words(self, keyword: str, file: str) -> bool:
+        with open(file) as f:
+            for line in f:
+                if keyword in line:
+                    return True
+        return False
```
Comment (EikanWang) [inline tools/test/test_codegen.py:649]: Please utilize `FileCheck`

File: tools/test/test_codegen.py [hunk 2]
```diff
@@ -498,6 +511,164 @@ def test_op_with_cpp_sig_generates_static_dispatch(self) -> None:
...
+
+    def test_generated_fils(self) -> None:
+        assert os.path.exists(os.path.join(self.install_dir, "ops/as_strided_native.h"))
```
Comment (EikanWang) [inline tools/test/test_codegen.py:652]: Please replace all assert with `self.assertXXX`

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Review by atalman — APPROVED** [review]
lgtm

File: torchgen/dest/register_dispatch_key.py [hunk 3]
```diff
@@ -62,6 +62,9 @@ def gen_registration_headers(
             headers.append("#include <ATen/cuda/EmptyTensor.h>")
     elif backend_index.dispatch_key == DispatchKey.MPS:
         headers.append("#include <ATen/mps/EmptyTensor.h>")
+    elif backend_index.dispatch_key == DispatchKey.XPU:
+        # XPU specific, this header resides in third_party/torch-xpu-ops
```
Comment (albanD) [inline torchgen/dest/register_dispatch_key.py:66]: This is not a real folder in this repo?

Comment (EikanWang) [reply torchgen/dest/register_dispatch_key.py:66]: @albanD , it is not a real folder. FYI: https://github.com/pytorch/pytorch/pull/120891#pullrequestreview-1929849651

**Comment by ZhiweiYan-96** [comment]
hi @albanD, great thanks for your comments. The header here resides in the `thirdy_party/torch-xpu-ops`, where most XPU operators are in it. The header is exposed to PyTorch build process by cmake change here https://github.com/pytorch/pytorch/pull/130082/files#diff-c5ee05f1e918772792ff6f2a3f579fc2f182e57b1709fd786ef6dc711fd68b27R1054.  So even, PyTorch main code base could not see these headers, these headers could be proper used during compilation.

**Comment by ZhiweiYan-96** [comment]
hi @albanD I may merge the change PR first. If you need me have further change, please give me your valuable suggestions, and I would append commit to refractor them, great thanks!

**Comment by EikanWang** [comment]
@ZhiweiYan-96 , please add document to describe the usage of this PR.

**Pull request closed without merging**